RFC: quiesce ranges [postponed]

dt opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8811)

<!-- Reviewable:end -->

dt committed:
rfc: quiesce ranges

jordanlewis commented:
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/quiesce_ranges.md, line 16 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPyU8nCXkfzkNqUCz0_:-KPyU8nCXkfzkNqUCz0a:be8ow7m) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L16)):*
> ```Markdown
> As clusters grow to potentially very large numbers of ranges, if access patterns
> are such that some of those ranges see no traffic, maintaining the raft groups
> and other state for them potentially wastes substantial resources for little
> ```

Consider dropping this second "potentially"

---

*[docs/RFCS/quiesce_ranges.md, line 59 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPyTE6X9CWeuFv7620A:-KPyTE6Y0ZAQVjGbOF8F:bc6qieo) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L59)):*
> ```Markdown
> 
> Once quiesced, any replica receiving a request should restart raft and trigger
> an election to be able to acquire the lease. Theortically, the lease holder
> ```

theoretically

---

*[docs/RFCS/quiesce_ranges.md, line 62 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPyUa6JVZUOzbDnnBio:-KPyUa6JVZUOzbDnnBip:bg3p43j) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L62)):*
> ```Markdown
> could serve reads for the remainder of its lease but if a request is received,
> it is reasonable to resume raft immediately as the range is no longer inactive
> and thus should no longer be quiesced.
> ```

What happens on the raft layer when a killed node is restarted? How does it figure out that a range is quiesced, and how does this interact with the lazy replica initialization? I think it could be useful to include these scenarios in the explanation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on docs/RFCS/quiesce_ranges.md:
```diff
@@ -0,0 +1,81 @@
+more than the election timeout, it was going to call the election anyway, thus
+ending the term that quiesced.
+
+Once quiesced, any replica receiving a request should restart raft and trigger
```
s/receiving a request/asked to propose a command/

tschottdorf commented on docs/RFCS/quiesce_ranges.md:
```diff
@@ -0,0 +1,81 @@
+ending the term that quiesced.
+
+Once quiesced, any replica receiving a request should restart raft and trigger
+an election to be able to acquire the lease. Theortically, the lease holder
```
s/Theort/Theor/

tschottdorf commented on docs/RFCS/quiesce_ranges.md:
```diff
@@ -0,0 +1,81 @@
+
+Once quiesced, any replica receiving a request should restart raft and trigger
+an election to be able to acquire the lease. Theortically, the lease holder
+could serve reads for the remainder of its lease but if a request is received,
```
This is already handled implicitly by the fact that we extend the lease well before it expires, so I don't think anything needs to be done here.

tschottdorf commented:
Consider also adding some discussion on the urgency of this proposal - I *think* it's fairly low compared to some other, more immediate problems we have with Raft performance: we bottleneck when many replicas make substantial progress simultaneously, in particular when snapshots are involved (https://github.com/cockroachdb/cockroach/issues/8638).
In that context, a back-of-the-envelope computation could also be interesting: if we have (say) a block writer table with single-point inserts going randomly all over the place, how often (on average) is each range hit by an insert in the long run? When does that time get long enough for quiescing to make sense (giving that the next insert will pay double)? At approximately what magnitude of number of Replicas do we expect the simple housekeeping (for example ticking them all) to be measurable? I think these are interesting questions - my hunch is that with coalesced heartbeats in, quiescence is really only useful or needed for very large data sets (which are correspondingly rare, at least in the foreseeable future), and that we *need* to allow for a lot of active Raft groups at a time in the common case because of the penalty associated with recreation.

tschottdorf commented on docs/RFCS/quiesce_ranges.md:
```diff
@@ -0,0 +1,81 @@
+
+This almost certainly needs to be implemented at least partially upstream in
+raft -- inspecting the follower state to determine when it is safe to stop
+heartbeating involves inspecting internal raft state.
```
Could add here that this feature may not be received enthusiastically in upstream Raft, as we're the only multi-range use case and the cost of letting a single Raft group live is marginal.

bdarnell commented on docs/RFCS/quiesce_ranges.md:
```diff
@@ -0,0 +1,81 @@
+raft -- inspecting the follower state to determine when it is safe to stop
+heartbeating involves inspecting internal raft state.
+
+Additionally, the first request to a range that has quiesced will see a
```
@spencerkimball has suggested handling quiescing in a different way, in which the former leader can optimistically attempt to revive the group with the same term and leader as before (if a node other than the former leader does the reviving, it will need to increment the term and have a full election).

bdarnell commented on docs/RFCS/quiesce_ranges.md:
```diff
@@ -0,0 +1,81 @@
+and thus should no longer be quiesced.
+
+# Drawbacks
+The largest drawback is reasoning about and maintaining the additional
```
And this complexity becomes moot when/if we implement node leases (#4288). Should we push forward with that plan instead of the stopgap solution of quiescent ranges?

bdarnell commented on docs/RFCS/quiesce_ranges.md:
```diff
@@ -0,0 +1,81 @@
+it is reasonable to resume raft immediately as the range is no longer inactive
+and thus should no longer be quiesced.
+
+# Drawbacks
```
Another drawback: This proposal makes assumptions about access patterns, namely that a cluster's access patterns are such that a substantial fraction of ranges could become quiescent. Some clusters will not see any benefit from this change.

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/quiesce_ranges.md, line 65 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPySungbcsYXCEHcZ5g-r1-65:-KQLTg-6A2uriiZOQf70:b-x57njj) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L65)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> And this complexity becomes moot when/if we implement node leases (#4288). Should we push forward with that plan instead of the stopgap solution of quiescent ranges?
</details>
Another alternative is to stop calling `RawNode.Tick()` and take over responsibility for replica related timers. @bdarnell I think you mentioned this somewhere else and my initial exploration of the raft code suggests it would be feasible. This approach would need to be folded in to some sort of coalesced heartbeat or node lease proposal to see any benefit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/quiesce_ranges.md, line 29 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KQLW1UDEkFx97YS0yHP:-KQLW1UEyRuZWyiGOk8W:bykh1dz) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L29)):*
> ```Markdown
> Additionally, some operations may be able to use the fact a range is in a
> quiesced state. For example, some possible implementations of bulk ingestion of
> *new* ranges could be simplified (by being able assume the raft state is
> ```

being able *to* assume

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811)*
<!-- Sent from Reviewable.io -->

dt commented:
Context for people who missed discussion with @mjibson and @tschotdorf last week:
While this was originally proposed to (in #357) purely for the network and resource savings, the ability to freeze raft state for a range is potentially useful e.g. for bulk ingestion.

So while the potential resource usage savings haven't yet seemed like enough, on their own, to make this a priority, its potential in building other things is why we wanted to explore it further now (though someone mentioning an idle cluster burning hundreds of gigs of network recently seems to suggest those savings might be worth *something* on their own too).

---

Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/quiesce_ranges.md, line 16 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPyU8nCXkfzkNqUCz0_:-KQREcWJeIxByjZHvU9b:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L16)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> Consider dropping this second "potentially"
</details>
Done.

---

*[docs/RFCS/quiesce_ranges.md, line 29 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KQLW1UDEkFx97YS0yHP:-KQREgKRtf2JfIZMoKki:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L29)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary>
> being able *to* assume
</details>
Done.

---

*[docs/RFCS/quiesce_ranges.md, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPySungbcsYXCEHcZ5g-r1-58:-KQREodlOUmivUUMxCkp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L58)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/receiving a request/asked to propose a command/
</details>
Done.

---

*[docs/RFCS/quiesce_ranges.md, line 59 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPySungbcsYXCEHcZ5g-r1-59:-KQRErhHiwuk6JJppVCS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L59)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> s/Theort/Theor/
</details>
Done.

---

*[docs/RFCS/quiesce_ranges.md, line 59 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPyTE6X9CWeuFv7620A:-KQREs4_zlsnyVMo9usJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L59)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> theoretically
</details>
Done.

---

*[docs/RFCS/quiesce_ranges.md, line 60 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPySungbcsYXCEHcZ5g-r1-60:-KQRN8WnLNhwCzACYDyL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L60)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> This is already handled implicitly by the fact that we extend the lease well before it expires, so I don't think anything needs to be done here.
</details>
Done.

---

*[docs/RFCS/quiesce_ranges.md, line 62 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPyUa6JVZUOzbDnnBio:-KQRVKWQghX79sJQCkrD:budij4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L62)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary>
> What happens on the raft layer when a killed node is restarted? How does it figure out that a range is quiesced, and how does this interact with the lazy replica initialization? I think it could be useful to include these scenarios in the explanation.
</details>
It'll just wake up everyone and then they'll all go back to sleep. Added a note about spurious wake ups being OK to drawbacks.

---

*[docs/RFCS/quiesce_ranges.md, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPySungbcsYXCEHcZ5g-r1-64:-KQROMd2Uoxn85JTrVoU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L64)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Another drawback: This proposal makes assumptions about access patterns, namely that a cluster's access patterns are such that a substantial fraction of ranges could become quiescent. Some clusters will not see any benefit from this change.
</details>
Done.

---

*[docs/RFCS/quiesce_ranges.md, line 65 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPySungbcsYXCEHcZ5g-r1-65:-KQRNmG48FePGm71VN39:b-ovsswl) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L65)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Another alternative is to stop calling `RawNode.Tick()` and take over responsibility for replica related timers. @bdarnell I think you mentioned this somewhere else and my initial exploration of the raft code suggests it would be feasible. This approach would need to be folded in to some sort of coalesced heartbeat or node lease proposal to see any benefit.
</details>
quiescing is also motivated by being able 

whether we discard the raft group or just stop ticking it, we still need to coordinate doing so across nodes at the same-ish time so that nodes don't think their peers have died

---

*[docs/RFCS/quiesce_ranges.md, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPySungbcsYXCEHcZ5g-r1-70:-KQRNlRrwkYcK2UmtJTu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L70)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Could add here that this feature may not be received enthusiastically in upstream Raft, as we're the only multi-range use case and the cost of letting a single Raft group live is marginal.
</details>
Done.

---

*[docs/RFCS/quiesce_ranges.md, line 72 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-KPySungbcsYXCEHcZ5g-r1-72:-KQRSF2-g404jli9g_dc:b-x5n501) ([raw file](https://github.com/cockroachdb/cockroach/blob/13d7205127a453635d5487dc975d77e51cb06fb9/docs/RFCS/quiesce_ranges.md#L72)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> @spencerkimball has suggested handling quiescing in a different way, in which the former leader can optimistically attempt to revive the group with the same term and leader as before (if a node other than the former leader does the reviving, it will need to increment the term and have a full election). 
</details>
Added under possible follow-on optimizations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-:-KQRq7uL9oCdUP4naGkm:b-qtwjn0)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
The document is written from the perspective that this is an optimization; if the primary motivation is now bulk ingestion, that should be reflected in the doc.

Quiescing-as-optimization and quiescing-for-bulk-ingestion have rather different needs. As discussed below, raft-level quiescing is fragile since there's no great way to synchronize it across all nodes and the slightest breeze will bring the raft group back. That's fine if it's an optimization, but not if you're semantically relying on it. I think for bulk ingestion it would be better to base things on the Freeze command and avoid low-level raft tinkering. 

---

Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-:-KQTT2z03ZCzoGDZK8Za:bt93o4u)*
<!-- Sent from Reviewable.io -->

dt commented:
To some degree I was hoping quiescing could be used by *both* bulk ingestion and as an resource usage optimization -- the machinery discussed herein for *how to quiesce* a range is hopefully general enough that it could be used either by  a "looks like this range is inactive so I can save some cycles" routine or by an explicit "i *need* this range to freeze its raft state so I can do something" bulk ingestion.

The current freeze doesn't quite work for some of the current proposals for the later, as they involve simple snapshots that happen to correspond to a an applied index in the future, and thus need to be sure raft is inactive and that index stays in the future -- but all of that I believe is covered in more detail in the forthcoming bulk-ingestion-of-ranges RFC that @paperstreet is writing -- this just got broken into a standalone RFC that that could reference and so we could gauge scope, particularly since it might have utility outside bulk ingestion as well.

---

Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-:-KQTUlRbeP5UUhR9Qzpf:b-pn9kj8)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Having discussed this a fair bit, I think that quiescing for restore purposes is a long shot (or at least so complicated that we're better off pondering alternatives), so I think that we're best off focusing here on the merit quiescing could have outside of that context.

---

Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8811#-:-KQTYfPRHvG3XI4_cXcm:b-2ibfaa)*
<!-- Sent from Reviewable.io -->

dt commented:
Ok, sounds like we've gathered more or less what we wanted to know for backup/restore -- quiescing active ranges seems tricky enough and the other benefits questionable enough that we might be better off exploring other ways for bulk ingestion first.

Should I change status to rejected and merge, or just close unmerged?

bdarnell commented:
Merge as rejected, or maybe as "postponed" or "deferred".

On Wed, Aug 31, 2016, 10:10 PM David Taylor <notifications@github.com>
wrote:

> Ok, sounds like we've gathered more or less what we wanted to know for
> backup/restore -- quiescing active ranges seems tricky enough and the other
> benefits questionable enough that we might be better off exploring other
> ways for bulk ingestion first.
>
> Should I change status to rejected and merge, or just close unmerged?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8811#issuecomment-243776307>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJzMmmgQtNy6qO6Lm_RWbPshWAcsQW8ks5qlYtegaJpZM4JseX->
> .
>

dt commented:
done.
